tower-http/src/auth/require_authorization_async.rs

```diff
@@ -0,0 +1,357 @@
+//! Authorize requests using the [`Authorization`] header asynchronously.
+//!
+//! [`Authorization`]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization
+//!
+//! # Example
+//!
+//! ```
+//! use tower_http::auth::{RequireAuthorizationAsyncLayer, AuthorizeRequestAsync};
+//! use hyper::{Request, Response, Body, Error};
+//! use http::{StatusCode, header::AUTHORIZATION};
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//! use futures_util::future::BoxFuture;
+//!
+//! #[derive(Clone, Copy)]
+//! struct MyAuth;
+//!
+//! impl AuthorizeRequestAsync for MyAuth {
+//!     type Output = UserId;
+//!     type Future = BoxFuture<'static, Option<UserId>>;
+//!     type ResponseBody = Body;
+//!
+//!     fn authorize<B>(&mut self, request: &Request<B>) -> Self::Future {
+//!         # Box::pin(async {
+//!             // ...
+//!             # None
+//!         })
+//!     }
+//!
+//!     fn on_authorized<B>(&mut self, request: &mut Request<B>, user_id: UserId) {
+//!         // Set `user_id` as a request extension so it can be accessed by other
+//!         // services down the stack.
+//!         request.extensions_mut().insert(user_id);
+//!     }
+//!
+//!     fn unauthorized_response<B>(&mut self, request: &Request<B>) -> Response<Body> {
+//!         Response::builder()
+//!             .status(StatusCode::UNAUTHORIZED)
+//!             .body(Body::empty())
+//!             .unwrap()
+//!     }
+//! }
+//!
+//! #[derive(Debug)]
+//! struct UserId(String);
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     // Access the `UserId` that was set in `on_authorized`. If `handle` gets called the
+//!     // request was authorized and `UserId` will be present.
+//!     let user_id = request
+//!         .extensions()
+//!         .get::<UserId>()
+//!         .expect("UserId will be there if request was authorized");
+//!
+//!     println!("request from {:?}", user_id);
+//!
+//!     Ok(Response::new(Body::empty()))
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service = ServiceBuilder::new()
+//!     // Authorize requests using `MyAuth`
+//!     .layer(RequireAuthorizationAsyncLayer::custom(MyAuth))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_core::ready;
+use http::{Request, Response};
+use http_body::Body;
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`RequireAuthorizationAsync`] which authorizes all requests using the
+/// [`Authorization`] header.
+///
+/// See the [module docs](crate::auth::require_authorization_async) for an example.
```

Comment:
I think we should `Async` a prefix. That way it matches `Filter` and `AsyncFilter` in tower.

```suggestion
/// See the [module docs](crate::auth::async_require_authorization) for an example.
```